File: scripts/test-caching.ts
/**
 * Test script to verify Claude prompt caching is working
 *
 * Run with: npx ts-node scripts/test-caching.ts
 *
 * Requirements for caching to work:
 * 1. Cached content must be at least 1024 tokens
 * 2. Same content must be sent in multiple requests
 * 3. Requests must occur within 5 minutes of cache creation
 */

import Anthropic from "@anthropic-ai/sdk";

// Initialize client with prompt caching beta header
const client = new Anthropic({
  defaultHeaders: {
    "anthropic-beta": "prompt-caching-2024-07-31",
  },
});

// Large static system prompt (needs to be >1024 tokens for caching)
const LARGE_SYSTEM_PROMPT = `You are Claude Code, an expert AI programming assistant helping a candidate during a technical interview assessment.

## Your Core Responsibilities

### 1. Code Assistance
- Help candidates understand problems and develop solutions
- Provide guidance on algorithms, data structures, and best practices
- Suggest improvements to code quality, efficiency, and readability
- Debug issues when tests fail
- Be concise but thorough in your explanations

### 2. Technical Guidance
- Explain complex concepts in simple terms
- Provide code snippets to illustrate ideas
- Suggest appropriate design patterns
- Help optimize performance where needed
- Guide candidates through debugging processes

### 3. Interview Support
- Act as a collaborative pair programming partner
- Encourage good software engineering practices
- Help candidates think through edge cases
- Provide hints without giving away solutions
- Support test-driven development approaches

## Important Guidelines

### What TO Do:
- Ask clarifying questions when requirements are unclear
- Encourage test-driven development
- Point out edge cases candidates should consider
- Provide code snippets to illustrate concepts
- Explain the reasoning behind suggestions
- Help debug issues systematically
- Suggest refactoring opportunities
- Recommend appropriate data structures
- Guide candidates through algorithm design
- Support incremental development

### What NOT To Do:
- Do NOT write the entire solution for them
- Do NOT reveal test case details or expected outputs
- Do NOT discuss candidate evaluation or scoring
- Do NOT compare candidates to others
- Do NOT reveal question difficulty levels
- Do NOT discuss the assessment algorithm
- Do NOT provide complete implementations without explanation
- Do NOT skip explanation of complex concepts
- Do NOT ignore code quality issues
- Do NOT dismiss candidate questions

## Communication Style

### Tone
- Be helpful, encouraging, and collaborative
- Maintain professionalism while being approachable
- Use clear and concise language
- Be patient with candidates who are struggling
- Celebrate progress and good approaches

### Explanations
- Break down complex problems into smaller steps
- Use analogies when helpful
- Provide context for recommendations
- Explain trade-offs between different approaches
- Reference relevant documentation when appropriate

### Code Examples
- Keep examples focused and relevant
- Include comments explaining key points
- Show both good and improved versions when suggesting changes
- Demonstrate idiomatic code patterns
- Include error handling in examples

## Technical Knowledge Areas

### Languages & Frameworks
- JavaScript/TypeScript: ES6+, Node.js, React, Vue, Angular
- Python: Django, Flask, FastAPI, data science libraries
- Go: Standard library, common patterns, concurrency
- Java: Spring, Maven/Gradle, JVM optimization
- Other languages as needed

### Best Practices
- SOLID principles
- Clean code guidelines
- Testing strategies (unit, integration, e2e)
- Code review best practices
- Documentation standards
- Performance optimization
- Security considerations
- Accessibility requirements

### Data Structures & Algorithms
- Arrays, linked lists, stacks, queues
- Trees, graphs, heaps
- Hash tables, sets, maps
- Sorting and searching algorithms
- Dynamic programming
- Graph algorithms
- String manipulation
- Time and space complexity analysis

## Assessment Integrity

Remember: Your role is to help candidates demonstrate their abilities, not to do the work for them. Guide them toward solutions while ensuring they understand the concepts and can apply them independently.

Be a helpful pair programming partner while maintaining assessment integrity. This is a learning experience for the candidate.

## Code Review Criteria

### Code Quality Factors
When reviewing code, consider these factors:
- Readability: Is the code easy to understand?
- Maintainability: Can this code be easily modified?
- Efficiency: Is the algorithm optimal for the use case?
- Correctness: Does the code handle all edge cases?
- Style: Does the code follow language conventions?

### Common Code Smells
Watch for these issues:
- Long methods that do too much
- Deeply nested conditionals
- Magic numbers without explanation
- Duplicated code blocks
- Poor variable naming
- Missing error handling
- Inconsistent formatting
- Unused variables or imports
- Overly complex expressions
- Missing documentation for public APIs

### Performance Considerations
Help candidates optimize for:
- Time complexity improvements
- Space complexity trade-offs
- Caching opportunities
- Lazy evaluation benefits
- Database query optimization
- Memory management
- Network request batching
- Asynchronous processing

## Testing Best Practices

### Unit Testing Guidelines
- Test one thing at a time
- Use descriptive test names
- Follow Arrange-Act-Assert pattern
- Mock external dependencies
- Test edge cases and error conditions
- Aim for high code coverage
- Keep tests independent
- Make tests deterministic

### Integration Testing
- Test component interactions
- Verify database operations
- Test API endpoints
- Check authentication flows
- Validate data transformations
- Test error propagation
- Verify logging and monitoring

## Security Considerations

### Common Vulnerabilities
Be aware of:
- SQL injection attacks
- Cross-site scripting (XSS)
- Cross-site request forgery (CSRF)
- Authentication bypasses
- Authorization failures
- Sensitive data exposure
- Insecure direct object references
- Security misconfiguration
- Using components with known vulnerabilities
- Insufficient logging and monitoring

### Secure Coding Practices
Recommend:
- Input validation and sanitization
- Parameterized queries
- Output encoding
- Proper authentication mechanisms
- Role-based access control
- Encryption for sensitive data
- Secure session management
- HTTPS everywhere
- Security headers
- Regular dependency updates

This comprehensive guide should help you assist candidates effectively while maintaining high standards for code quality and security.`;

interface CacheMetrics {
  inputTokens: number;
  outputTokens: number;
  cacheCreationInputTokens: number;
  cacheReadInputTokens: number;
}

async function testCaching(): Promise<void> {
  console.log("🧪 Testing Claude Prompt Caching\n");
  console.log("=" .repeat(50));

  const metrics: CacheMetrics[] = [];

  // Test messages that change each request
  const testQuestions = [
    "What is a binary search tree?",
    "How do I implement a linked list?",
    "Explain the difference between BFS and DFS.",
    "What is memoization?",
    "How does a hash table work?",
  ];

  // Build system prompt with cache_control
  const systemBlocks = [
    {
      type: "text" as const,
      text: LARGE_SYSTEM_PROMPT,
      cache_control: { type: "ephemeral" as const },
    },
  ];

  console.log(`\nSystem prompt length: ~${Math.round(LARGE_SYSTEM_PROMPT.length / 4)} tokens (estimated)`);
  console.log("Minimum for caching: 1024 tokens\n");

  for (let i = 0; i < testQuestions.length; i++) {
    const question = testQuestions[i];
    console.log(`\n📨 Request ${i + 1}: "${question.substring(0, 40)}..."`);

    const startTime = Date.now();

    const response = await client.messages.create({
      model: "claude-sonnet-4-5-20250929",
      max_tokens: 100,
      system: systemBlocks as any,
      messages: [
        { role: "user", content: question },
      ],
    });

    const elapsed = Date.now() - startTime;

    // Extract usage including cache metrics
    const usage = response.usage as any;
    const cacheCreation = usage.cache_creation_input_tokens || 0;
    const cacheRead = usage.cache_read_input_tokens || 0;

    metrics.push({
      inputTokens: usage.input_tokens,
      outputTokens: usage.output_tokens,
      cacheCreationInputTokens: cacheCreation,
      cacheReadInputTokens: cacheRead,
    });

    console.log(`   ⏱️  Time: ${elapsed}ms`);
    console.log(`   📥 Input tokens: ${usage.input_tokens}`);
    console.log(`   📤 Output tokens: ${usage.output_tokens}`);
    console.log(`   💾 Cache created: ${cacheCreation} tokens`);
    console.log(`   ✅ Cache read: ${cacheRead} tokens`);

    if (cacheRead > 0) {
      console.log(`   🎯 CACHE HIT! Saved ~${Math.round(cacheRead * 0.9)} tokens worth of cost`);
    } else if (cacheCreation > 0) {
      console.log(`   📝 Cache created for future requests`);
    } else {
      console.log(`   ⚠️  No caching detected`);
    }

    // Small delay between requests
    await new Promise(resolve => setTimeout(resolve, 500));
  }

  // Summary
  console.log("\n" + "=".repeat(50));
  console.log("📊 CACHING SUMMARY\n");

  const totalCacheCreated = metrics.reduce((sum, m) => sum + m.cacheCreationInputTokens, 0);
  const totalCacheRead = metrics.reduce((sum, m) => sum + m.cacheReadInputTokens, 0);
  const totalInput = metrics.reduce((sum, m) => sum + m.inputTokens, 0);

  console.log(`Total requests: ${metrics.length}`);
  console.log(`Total input tokens: ${totalInput}`);
  console.log(`Cache created tokens: ${totalCacheCreated}`);
  console.log(`Cache read tokens: ${totalCacheRead}`);

  if (totalCacheRead > 0) {
    const savingsPercent = Math.round((totalCacheRead / totalInput) * 100);
    console.log(`\n✅ CACHING IS WORKING!`);
    console.log(`   Cache hit rate: ${savingsPercent}% of input tokens from cache`);
    console.log(`   Estimated savings: ~$${((totalCacheRead * 0.9 * 3) / 1_000_000).toFixed(4)} (at $3/MTok)`);
  } else if (totalCacheCreated > 0) {
    console.log(`\n⚠️  Cache was created but not read`);
    console.log(`   This may happen if:`);
    console.log(`   - Requests are too far apart (>5 min TTL)`);
    console.log(`   - System prompt changed between requests`);
  } else {
    console.log(`\n❌ NO CACHING DETECTED`);
    console.log(`   Possible reasons:`);
    console.log(`   - System prompt is under 1024 tokens`);
    console.log(`   - cache_control not properly set`);
    console.log(`   - API doesn't support caching for this model`);
  }
}

// Run the test
testCaching()
  .then(() => {
    console.log("\n✨ Test complete!");
    process.exit(0);
  })
  .catch((error) => {
    console.error("\n❌ Test failed:", error);
    process.exit(1);
  });

File: langgraph-agents/test_caching.py
#!/usr/bin/env python3
"""
Quick test script to verify Anthropic prompt caching is working correctly.

This script makes two identical requests and checks if the second one uses cached tokens.
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from langchain_anthropic import ChatAnthropic
from langchain_core.messages import HumanMessage, SystemMessage

from config import settings

LONG_SYSTEM_PROMPT = """You are a senior software engineer helping candidates during technical interviews.

You have extensive knowledge of:
- Data structures and algorithms
- System design and architecture
- Programming languages: Python, JavaScript, TypeScript, Go, Rust, Java
- Best practices for clean, maintainable code
- Testing and debugging strategies

Your role is to:
1. Help candidates understand problems
2. Guide them through solutions without giving away answers directly
3. Review their code and provide constructive feedback
4. Explain concepts when needed

Remember to be encouraging but honest in your feedback.
""" * 10  # Repeat to ensure it's long enough for caching (min ~1024 tokens)


async def test_caching():
    print("=" * 60)
    print("Anthropic Prompt Caching Test")
    print("=" * 60)
    print(f"\nModel: {settings.coding_agent_model}")
    print(f"Caching Enabled: {settings.enable_prompt_caching}")
    print(f"System Prompt Length: {len(LONG_SYSTEM_PROMPT)} chars")

    # Create LLM with caching headers
    extra_headers = {}
    if settings.enable_prompt_caching:
        extra_headers["anthropic-beta"] = "prompt-caching-2024-07-31"
        print("Beta Header: anthropic-beta=prompt-caching-2024-07-31")

    llm = ChatAnthropic(
        model=settings.coding_agent_model,
        max_tokens=256,
        api_key=settings.anthropic_api_key,
        extra_headers=extra_headers if extra_headers else None,
    )

    # Create system message with cache_control in content block
    if settings.enable_prompt_caching:
        system_msg = SystemMessage(content=[
            {
                "type": "text",
                "text": LONG_SYSTEM_PROMPT,
                "cache_control": {"type": "ephemeral"}
            }
        ])
        print("System Message: Using content block with cache_control")
    else:
        system_msg = SystemMessage(content=LONG_SYSTEM_PROMPT)
        print("System Message: Plain content (no caching)")

    # First request (should create cache)
    print("\n" + "-" * 40)
    print("Request 1 (should CREATE cache)")
    print("-" * 40)

    messages1 = [
        system_msg,
        HumanMessage(content="What is a linked list?")
    ]

    response1 = await llm.ainvoke(messages1)

    # Extract usage metadata
    usage1 = getattr(response1, 'usage_metadata', {})
    response_meta1 = getattr(response1, 'response_metadata', {})

    print(f"\nResponse: {response1.content[:100]}...")
    print(f"\nUsage Metadata: {usage1}")
    print(f"Response Metadata keys: {list(response_meta1.keys())}")

    # Check for cache metrics in response
    cache_creation_1 = response_meta1.get('usage', {}).get('cache_creation_input_tokens', 0)
    cache_read_1 = response_meta1.get('usage', {}).get('cache_read_input_tokens', 0)
    input_tokens_1 = response_meta1.get('usage', {}).get('input_tokens', 0)

    print(f"\nCache Creation Tokens: {cache_creation_1}")
    print(f"Cache Read Tokens: {cache_read_1}")
    print(f"Input Tokens: {input_tokens_1}")

    if cache_creation_1 > 0:
        print("✓ Cache CREATED successfully!")
    elif cache_read_1 > 0:
        print("! Cache already existed (read from cache)")
    else:
        print("✗ No cache activity detected")

    # Wait a moment
    await asyncio.sleep(1)

    # Second request (should read from cache)
    print("\n" + "-" * 40)
    print("Request 2 (should READ from cache)")
    print("-" * 40)

    messages2 = [
        system_msg,
        HumanMessage(content="What is a binary tree?")
    ]

    response2 = await llm.ainvoke(messages2)

    # Extract usage metadata
    usage2 = getattr(response2, 'usage_metadata', {})
    response_meta2 = getattr(response2, 'response_metadata', {})

    print(f"\nResponse: {response2.content[:100]}...")
    print(f"\nUsage Metadata: {usage2}")

    cache_creation_2 = response_meta2.get('usage', {}).get('cache_creation_input_tokens', 0)
    cache_read_2 = response_meta2.get('usage', {}).get('cache_read_input_tokens', 0)
    input_tokens_2 = response_meta2.get('usage', {}).get('input_tokens', 0)

    print(f"\nCache Creation Tokens: {cache_creation_2}")
    print(f"Cache Read Tokens: {cache_read_2}")
    print(f"Input Tokens: {input_tokens_2}")

    if cache_read_2 > 0:
        print("✓ Cache HIT! Reading from cache.")
        savings = cache_read_2 * 0.9  # 90% savings on cached tokens
        print(f"✓ Estimated savings: {savings:.0f} tokens worth")
    elif cache_creation_2 > 0:
        print("! Cache miss - created new cache (unexpected)")
    else:
        print("✗ No cache activity detected")

    # Summary
    print("\n" + "=" * 60)
    print("SUMMARY")
    print("=" * 60)

    if cache_creation_1 > 0 and cache_read_2 > 0:
        print("\n✓ CACHING IS WORKING CORRECTLY!")
        print(f"  - Request 1: Created cache ({cache_creation_1} tokens)")
        print(f"  - Request 2: Cache hit ({cache_read_2} tokens)")
    elif cache_read_1 > 0 and cache_read_2 > 0:
        print("\n✓ CACHING IS WORKING (cache already existed)")
        print("  - Both requests used cached tokens")
    elif not settings.enable_prompt_caching:
        print("\n! Caching is DISABLED in settings")
        print("  - Set enable_prompt_caching=True to enable")
    else:
        print("\n✗ CACHING MAY NOT BE WORKING")
        print("  Possible causes:")
        print("  1. System prompt too short (needs ~1024+ tokens)")
        print("  2. Beta header not being sent correctly")
        print("  3. cache_control not in content block format")
        print("\n  Debug info:")
        print(f"  - Response metadata: {response_meta2}")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    asyncio.run(test_caching())
